lib/src/rules/library_annotations.dart

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/dart/element/extensions.dart';
+import 'package:meta/meta_meta.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Attach library annotations to library directives.';
+
+const _details = r'''
+Attach library annotations to library directives, rather than
+some other library-level element.
+
+**BAD:**
+```dart
+import 'package:test/test.dart';
+
+@TestOn('browser')
+void main() {}
+```
+
+**GOOD:**
+```dart
+@TestOn('browser')
+library;
+import 'package:test/test.dart';
+
+void main() {}
+```
+
+**NOTE:** An unnamed library, like `library;` above, is only supported in Dart
+2.19 and later. Code which might run in earlier versions of Dart will need to
+provide a name in the `library` directive.
+''';
+
+class LibraryAnnotations extends LintRule {
+  static const LintCode code = LintCode('library_annotations',
+      'This annotation must be attached to a library directive.',
+      correctionMessage:
+          'Try attaching library annotations to library directives.');
+
+  LibraryAnnotations()
+      : super(
+            name: 'library_annotations',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  LintCode get lintCode => code;
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LibraryAnnotations rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    for (var directive in node.directives) {
+      if (directive is PartDirective) {
+        return;
+      }
+      if (directive is! LibraryDirective) {
+        _check(directive);
+      }
+    }
+
+    node.declarations.forEach(_check);
+  }
+
+  void _check(AnnotatedNode node) {
+    for (var annotation in node.metadata) {
+      var elementAnnotation = annotation.elementAnnotation;
+      if (elementAnnotation == null) {
+        return;
+      }
+
+      if (elementAnnotation.targetKinds.length == 1 &&
+          elementAnnotation.targetKinds.contains(TargetKind.library) &&
+          (node.parent as CompilationUnit?)?.directives.first == node) {
```

Comment:
True, but if there are no directives, then no annotation which has `TargetKind.library` can exist in the file. In any case, I've fixed to be defensive.

> It looks like the code is ignoring library directives everywhere except on the first directive. Should it also check the first declaration in the case where there are no directives?

Sorry, I don't follow. Can you explain more?

Reply:
> ... if there are no directives, then no annotation which has `TargetKind.library` can exist in the file.

Oh, I always thought that we allowed library annotations / comments on the first declaration when there were no directives. Of course, we're heading away from allowing them anywhere other than on a `library` directive, so it's probably not a big deal.

> Sorry, I don't follow. Can you explain more?

That might be because of a typo. It should have read

> It looks like the code is ignoring library-specific annotations everywhere except on the first directive.

That comment was also predicated on the assumption that they were accepted on the first declaration, so if that's not the case then the comment is moot.

Reply:
The exception in best_practices_verifier only allows a `TargetKind.library` annotation to be placed on the first directive, not on the first declaration:

https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/lib/src/error/best_practices_verifier.dart#L1749

So I mirrored that here.